### PR TITLE
added frame accurate seeking demo with SMPTE timecodes

### PR DESCRIPTION
**Pull request by bitmovin-kenny**
Showcase for the player handling seeking with SMPTE Timecodes,

The FrameAccurateControls.ts offers functions to load a new asset and to get the current SMPTE timestamp as well as to seek to a given SMPTE timestamp.

### CODE REVIEW

File: frameaccurate/index.html [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        <div class="clearfix"></div>
+    </div>
+
+    <div class="container-fluid">
+        <div class="content">
+            <div class="row">
+                <div id="test-asset-wrapper" class="col-lg-12 text-center">
+                    <div>
+                        <h3>Test assets:</h3>
+                        <select id="test-assets"></select>
```
Comment (dweinber) [inline frameaccurate/index.html]: Please select one asset always and remove the selection for the sake of simplicity.

File: frameaccurate/js/frameaccurate-demo.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
+const assets = [
+  /**
+   * Add your own assets here
+   */
+  {
+    name: '24 FPS, 1 frame off, some duplicate frames',
```
Comment (dweinber) [inline frameaccurate/js/frameaccurate-demo.js]: Please remove the frame off & duplicate frame infos for all assets.

File: README.md [hunk 3]
```diff
@@ -16,6 +16,8 @@ This showcases are build around the Bitmovin Adaptive Streaming Player, demonstr
     * [ON_DOWNLOAD_FINISHED](events/onDownloadFinished.html): Report download errors for live streams using the ON_DOWNLOAD_FINISHED player event.
     * [eventConstants](events/eventConstants.html): Use event constants for registering event handlers at the player.
     * [ON_SEGMENT_REQUEST_FINISHED](events/onSegmentRequestFinished.html): Use ON_SEGMENT_REQUEST_FINISHED event to analyze which segment has been downloaded and implement custom workflows according to the HTTP status code.
+* [**frameaccurate**](frameaccurate/)
+    * [Frame Accurate Control](frameaccurate/js/FrameAccurateControls.ts): Bitmoinv Player Wrapper for SMTPE timestamp seeking + stepping
```
Comment (dweinber) [inline README.md]: We usually link the demos here

Comment (bitmovin-kenny) [reply README.md]: The other links i've seen link to demos which consist of a single html file, i thought the SMTPE Controller would be more interesting here

File: frameaccurate/js/frameaccurate-demo.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+window.setInterval(function(){
+  let currentTime = smtpeController.getCurrentSmpte();
+  // let currentTime = toSmtpeTimeCode(player && player.getCurrentTime() || 0, false, true, true);
+  // only update on change so you can copy paste it
+  if (currentTimeDisplay.innerHTML !== String(currentTime)) {
+    currentTimeDisplay.innerHTML = currentTime;
+  }
+}, 250);
+
+  function step(stepSize) {
```
Comment (HalfbyteHeroes) [inline frameaccurate/js/frameaccurate-demo.js]: Wrong indentation 😞

File: README.md (hunk 3 above)
Comment (dweinber) [reply README.md]: Makes sense, but I think the entry point of a demo is still the best option, to see which files are linked and so on. It's easy to navigate to all the included JS files, but it's much harder to go the other way around.

File: frameaccurate/js/frameaccurate-demo.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+let stepForwardButton    = null;
+let seekButton           = null;
+let smtpeSeekTime        = null;
+let testAssetSelection   = null;
+let testAssetDescription = null;
+
+let smtpeController = null;
+
+let conf = {
+  key: '89f6ed6c-ab0e-46c2-ac47-5665e60c3c41',
```
Comment (dweinber) [inline frameaccurate/js/frameaccurate-demo.js]: Should be `'YOUR KEY HERE'` 🙁

File: frameaccurate/js/frameaccurate-demo.js [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+  {
+    name: '24 FPS',
+    source: {
+      progressive: '//bitdash-a.akamaihd.net/content/art-of-motion-frame-numbers/artofmotion_24fps_framenumbers.mp4'
+    },
+    frameRate: 24,
+  },
+];
+
+let player               = null;
```
Comment (dweinber) [inline frameaccurate/js/frameaccurate-demo.js]: Code style: please don't align.

File: frameaccurate/js/FrameAccurateControls.ts [hunk 7]
```diff
@@ -0,0 +1,255 @@
+
+class FrameAccurateControls {
+
+  constructor(private player: any, private assetDescription: AssetDescription) {
```
Comment (dweinber) [inline frameaccurate/js/FrameAccurateControls.ts]: I'd prefer putting the variable explicitly in the class (even though it technically doesn't make a difference)

File: frameaccurate/js/FrameAccurateControls.ts [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+  }
+
+  public load(asset: AssetDescription) {
```
Comment (dweinber) [inline frameaccurate/js/FrameAccurateControls.ts]: Return type

File: frameaccurate/js/FrameAccurateControls.ts [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+      throw error;
+    }
+  }
+
+  public getCurrentSmpte(): string {
+    const currentTime = this.player.getCurrentTime();
+    return SmpteTimestamp.fromTimeWithAdjustments(currentTime, this.assetDescription).toString();
+  }
+
+  public step(stepSize: number) {
```
Comment (dweinber) [inline frameaccurate/js/FrameAccurateControls.ts]: Return type

File: frameaccurate/js/FrameAccurateControls.ts [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+      smpte.addFrame(stepSize, true);
+    }
+    this.seekToSMPTE(smpte.toString());
+  }
+}
+
+class AssetDescription {
+  public frameDuration: number;
+  public offsetToMidFrame: number;
+  constructor(public name: string, public sourceConfig: SourceConfig, public framesPerSecond: number,
```
Comment (dweinber) [inline frameaccurate/js/FrameAccurateControls.ts]: Add new line before the constructor.

File: frameaccurate/js/FrameAccurateControls.ts [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+              public adjustmentFactor?: number, public framesDroppedAtFullMinute?: number) {
+    if (adjustmentFactor == null) {
+      this.adjustmentFactor = Math.ceil(framesPerSecond) / framesPerSecond;
+    }
+    if(framesDroppedAtFullMinute == null) {
+      if (framesPerSecond === 29.97 || framesPerSecond === 29.98) {
```
Comment (dweinber) [inline frameaccurate/js/FrameAccurateControls.ts]: But what if it's e.g. `29.97003`?

File: frameaccurate/js/FrameAccurateControls.ts [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+
+class SmpteTimestamp {
+  public frame: number;
+  public seconds: number;
+  public minutes: number;
+  public hours: number;
+
+  public constructor(smtpeTimestamp: string, private assetDescription: AssetDescription) {
+    if (smtpeTimestamp && SmpteTimestamp.validateTimeStamp(smtpeTimestamp, assetDescription.framesPerSecond)) {
+
```
Comment (dweinber) [inline frameaccurate/js/FrameAccurateControls.ts]: Unnecessary new line

File: frameaccurate/js/FrameAccurateControls.ts [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+    } else {
+      this.hours = 0;
+      this.minutes = 0;
+      this.seconds = 0;
+      this.frame = 0;
+    }
+  }
+
+  public static validateTimeStamp(smtpeTimestamp: string, framesPerSecond: number): boolean {
+
```
Comment (dweinber) [inline frameaccurate/js/FrameAccurateControls.ts]: Unnecessary new line

File: frameaccurate/js/FrameAccurateControls.ts [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+    // validate SMTPE timecode
+    let isValidSMPTETimeCode = new RegExp(/(^(?:(?:[0-1][0-9]|[0-2][0-3]):)(?:[0-5][0-9]:){2}(?:[0-6][0-9])$)/);
```
Comment (dweinber) [inline frameaccurate/js/FrameAccurateControls.ts]: From [Wikipedia](https://en.wikipedia.org/wiki/SMPTE_timecode#Drop_frame_timecode): 
> While non-drop timecode is displayed with colons separating the digit pairs—"HH:MM:SS:FF"—drop frame is usually represented with a semi-colon (;) or period (.) as the divider between all the digit pairs—"HH;MM;SS;FF", "HH.MM.SS.FF"—or just between the seconds and frames—"HH:MM:SS;FF" or "HH:MM:SS.FF". The period is usually used on VTRs and other devices that don't have the ability to display a semi-colon.
>
> Drop frame timecode is typically abbreviated as DF and non-drop as NDF.

Is this taken into account? We should use the `;` only between seconds and frames for this case.

Comment (bitmovin-kenny) [reply frameaccurate/js/FrameAccurateControls.ts]: This is not taken into account, but would be not be too hard to implement

File: frameaccurate/js/FrameAccurateControls.ts [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+      smtpe.addFrame(framesToAdd, false);
+      if (smtpe.minutes % 10 !== 0 && minutesBefore !== smtpe.minutes) {
+        smtpe.addFrame(assetDesc.framesDroppedAtFullMinute, false);
+      }
+    }
+
+    return smtpe;
+  }
+
+  public addFrame(framesToAdd: number, fixFrameHoles: boolean = true) {
```
Comment (dweinber) [inline frameaccurate/js/FrameAccurateControls.ts]: Please check return values for all methods.

**Pull request merged**